# Machine Learning and Statistics


## Tasks 2020








## Student: Doris Zdravkovic 

_______________________________________________________________________________________________________________________________

#### Task 1.  Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places.

Completion of the task:


1. Research the topic by repeated watching the lecture video and find more information online.
2. Based on information that I find I will write a python function that calculates square root of 2 and use Pyhton to set the output to 100 decimal places without using any libraries.
3. Test my function using Python library.
4. Write conclusion and references.

#### 1.1. Square root of 2

The square root of 2 or root 2 is represented using the square root symbol √ and written as √2 whose value is 1.414. This value is widely used in mathematics. It is an irrational number as it cannot be expressed as a fraction and has an infinite number of decimals. So, the exact value of the root of 2 cannot be determined. The square root of 2 is the number which when multiplied with itself gives the result as 2. The root 2 value can be computed to 10 trillion digits. For general use, its value is truncated and is used as 1.414 to make calculations easy. 


To find the square root value of a number, we usually check if the number is a perfect square or not. It is easy to find the root of perfect squares, but for non-perfect squares, we have to use the long division method to find the root value. Numbers like 2, 3, 5, 20, etc. are non-perfect squares, but 4, 9, 25, etc. are perfect squares, which gives the whole number when we take the root of them (byjus.com).

#### 1.2. Newton's method

In numerical analysis, Newton's method, also known as the Newton–Raphson method, named after Isaac Newton and Joseph Raphson, is a root-finding algorithm which produces successively better approximations to the roots (or zeroes) of a real-valued function. The root of a function is the point at which f(x)=0. Many equations have more than one root. Every real polynomial of odd degree has an odd number of real roots. Newton-Raphson is an iterative method that begins with an initial guess of the root. The method uses the derivative of the function f′(x) as well as the original function f(x), and thus only works when the derivative can be determined (aaronschlegel.me)
.
In the next paragraph each step of Newton's method will be explained. 


###### Algorithm

This algorithm describes steps or phases of the method. I find these steps very useful for better understanding of the function that will be of a great help when actually doing Pyhton coding of root 2 (hackernoon.com).

1. Take a reasonable guess (approximate root) for the square root.
2. Add the approximate root with the original number divided by the approximate root and divide by 2.
x_i := (x_i + n / x_i) / 2
3. Continue step 2 until the difference in the approximate root along the iterations is less than the desired value (or precision value).
4. The approximate root is the square root we want.



My Python code is based on the algorithm below. The algorithm was taken from the above mentioned web page and there will be some adjustments done in order to complete the task as necessary. Firstly, function is defined as mySqrt and x is a number that we want to find the root of. In this case x will be 2. The variable r is equal to x and used as the guess of starting square root which means that r would be eqal to 2. This part will be changed in my Pyhton code because I will set guess variable to 1 as a reasonable guess or approximate root. As there is no best initial guess according to computingskillset.com I will set my guess variable to 1. After the guess we move to next step and set the precision. Next step is to loop the root as long as it is higher than precision and it has to be divided by 2. 

def mySqrt(x):

    r = x
    precision = 10 ** (-10)
    
    while abs(x - r * r) > precision:
        r = (r + x / r) / 2
        
    return r
    
    
As I already mentioned this function will be used as layout for my code and there will be some changes made. 

In [86]:
# Newton's square root of 2
def sqrt2(x = 2):
   # x = 2 because we are calculating root of 2
    # Reasonable guess for the square root of 2.
    guess = 1
    # As per code above
    precision = 10 ** (-10)
# Iteration is set to 0
    iter = 0
    
# Loop guess*guess until root of 2 is reached

    while abs(x - guess * guess) > precision:
# Calcuation of even more precise guess for the square root of 2.
        guess = (guess+x/guess) / 2 
#  Number of iterations.
        iter += 1

# Returning root of 2 and number of iterations.
    return guess, iter
 
# First element = sqrt(2); second = number of iterations.
# Getting the root of 2.
root = sqrt2()[0]

# Getting the number of iterations.
n = sqrt2()[1]
# type(n)

# Printing the root of 2  
print(root)

# Using string formatting to print number of decimal places required.
print("Number of iterations:", n)
# Setting root of 2 to 100 decimal places.
format(root, '.100f')   

1.4142135623746899
Number of iterations: 4


'1.4142135623746898698271934335934929549694061279296875000000000000000000000000000000000000000000000000'

I was curious if I would get the same result if I changed the guess variable. So I will test the same code again with guess variable changed to 1.33 instead of 1. I have randomly chosen number 1.33. From the code below we can see that there are slight differences. As my guess variable was higher than in the first code (1.33 instead of 1) the number of iterations decreased from 4 to 3. The root of 2 has slightly changed as well. The only thing that remained the same are zeros at the end of the root of 2. 

In [87]:
# Newton's square root of 2
def sqrt2(x = 2):
   # x = 2 because we are calculating root of 2
    # Reasonable guess for the square root of 2.
    guess = 1.33
    # As per code above
    precision = 10 ** (-10)
# Iteration is set to 0
    iter = 0
    
# Loop guess*guess until root of 2 is reached

    while abs(x - guess * guess) > precision:
# Calcuation of even more precise guess for the square root of 2.
        guess = (guess+x/guess) / 2 
#  Number of iterations.
        iter += 1

# Returning root of 2 and number of iterations.
    return guess, iter
 
# First element = sqrt(2); second = number of iterations.
# Getting the root of 2.
root = sqrt2()[0]

# Getting the number of iterations.
n = sqrt2()[1]
# type(n)

# Printing the root of 2  
print(root)

# Using string formatting to print number of decimal places required.
print("Number of iterations:", n)
# Setting root of 2 to 100 decimal places.
format(root, '.100f')   

1.4142135623753198
Number of iterations: 3


'1.4142135623753198103713657474145293235778808593750000000000000000000000000000000000000000000000000000'

#### 1.3. Testing the square root 2

Since I have tested my code twice with different guess variable and the end result was different, I will use 2 different Python modules to test if the square root of 2 shows any similar results. Python has a built-in module that you can use for mathematical tasks called math.sqrt and the decimal module that provides support for decimal floating point arithmetic (docs.python.org).

In [88]:
# Code taken from: 
# https://docs.python.org/2/library/decimal.html

# Import decimal module
from decimal import *
# Formating to 100 decimal places
getcontext().prec = 100
# Root of 2
Decimal(2).sqrt()

Decimal('1.414213562373095048801688724209698078569671875376948073176679737990732478462107038850387534327641573')

In [89]:
# Code adjusted from: https://www.geeksforgeeks.org/python-math-function-sqrt/
# Import the math module  
import math  
  
# Print the square root of  2  
print(math.sqrt(2)) 
# Format to 100 decimal places
format(sqr,'.100f' )

1.4142135623730951


'1.4142135623730951454746218587388284504413604736328125000000000000000000000000000000000000000000000000'

Interestingly, all 4 Python codes that I have run gave slightly different result. Even modules "decimal" and "math" have output dissimilar result which was unexpected. Before I ran the test I was convinced that the output of the two modules would be the same or at least very similar as they are both Python libraries made by professionals or there is some kind of generalized, standard algorithm used to calculate root of any number. 

##### 1.4. Conclusion

After doing all these tests on root 2 I have realized why this topic is interesting for research. I will be going back at what I have written at the beginning of the project when I was trying to explain the importance of root 2. It says that "the exact value of the root of 2 cannot be determined" and even if I didn't take that information seriously at first, now after four different testings I can confirm the same as I got 4 different results. What matters most is that all results output what we call root of 2, and that is a float number 1.414, therefore we can say that all four algorithms have been sucessful. The most precise one was algorithm using decimal module as it is the only one that does not end with numerous zeros. It is not repetitive as other results. The reason might be the number of decimal points. As the number of decimal points increases the algorithm is not able to produce the precise result so it outputs zeros instead. What is interesting about this project is even though decimal module gave us "correct" results, without outputting zeros, we still can't be sure if that is the correct answer as the exact value of root of 2 cannot be precisely determined. 

##### References

Square root of 2: https://byjus.com/maths/square-root-of-2/

Newton's Method for Finding Equation Roots: https://aaronschlegel.me/newtons-method-equation-roots.html

Calculating the Square Root of a Number using the  Newton-Raphson Method: https://hackernoon.com/calculating-the-square-root-of-a-number-using-the-newton-raphson-method-a-how-to-guide-yr4e32zo

How to Find the Initial Guess in Newton’s Method: https://computingskillset.com/solving-equations/how-to-find-the-initial-guess-in-newtons-method/ 

Decimal fixed point and floating point arithmetic: https://docs.python.org/2/library/decimal.html

Python math function | sqrt(): https://www.geeksforgeeks.org/python-math-function-sqrt/

Extra readings:

Square root of 2: https://en.wikipedia.org/wiki/Square_root_of_2

The square root of 2: https://cosmosmagazine.com/mathematics/the-square-root-of-2/

Proof that the square root of 2 is irrational: https://www.khanacademy.org/math/algebra/x2f8bb11595b61c86:irrational-numbers/x2f8bb11595b61c86:proofs-concerning-irrational-numbers/v/proof-that-square-root-of-2-is-irrational

The Newton-Raphson Method: http://www.sosmath.com/calculus/diff/der07/der07.html

#### Task 2. Use scipy.stats to verify given value and calculate the associated p value of Chi squared test.

Completion of the task:


1. Research chi squared test online.
2. Based on information that I find I will write a chi square python function in scipy.stats to verify given Chi square value and calculate p value. 
3. Write analysis and conclusion
4. Write references.

#####  2.1 Chi squared test


The Chi Square statistic is commonly used for testing relationships between categorical variables.  The null hypothesis of the Chi-Square test is that no relationship exists on the categorical variables in the population; they are independent. The Chi-Square statistic is most commonly used to evaluate Tests of Independence when using a crosstabulation (also known as a bivariate table).  Crosstabulation presents the distributions of two categorical variables simultaneously, with the intersections of the categories of the variables appearing in the cells of the table.  The Test of Independence assesses whether an association exists between the two variables by comparing the observed pattern of responses in the cells to the pattern that would be expected if the variables were truly independent of each other.  Calculating the Chi-Square statistic and comparing it against a critical value from the Chi-Square distribution allows the researcher to assess whether the observed cell counts are significantly different from the expected cell counts [1]. 



Example chi-squared test for categorical data

Suppose there is a city of 1,000,000 residents with four neighborhoods: A, B, C, and D. A random sample of 650 residents of the city is taken and their occupation is recorded as "white collar", "blue collar", or "no collar". The null hypothesis is that each person's neighborhood of residence is independent of the person's occupational classification.The data are tabulated as:

![alt text](tablica_wiki.png)

What we see above is the example of chi-squared test for categorical data that is the base for my project. The aim of the project is to verify if chi sqared test using Python library scipy.stats will give me the same result as the one that is shown in this example from wikipedia. Calculations on wikipedia show that the Chi-sqared test in this particular case with these data is 24.6. 

##### 2.2 Calculation of Chi sqared test and p-value

In [1]:
# Calculating chi square using scipy.stats
import scipy.stats as ss
# Plotting
import matplotlib.pyplot as plt

In [3]:
# Creating the contingency table for our example (neighbourhoods)
white_collar = [90, 60, 104, 95] # Data for white collar
blue_collar = [30, 50, 51, 20] # Data for blue collar
no_collar = [30, 40, 45, 35] # Data for no collar
total = [white_collar, blue_collar, no_collar]
print(total)

[[90, 60, 104, 95], [30, 50, 51, 20], [30, 40, 45, 35]]


More about Chi squared statistics:

- chi2_contingency: this function takes list of numbers (neighbourhood data) and calculates statistics, p-value, degrees of freedom and expected frequencies. 

- p-value:  is a measure of the probability that an observed difference could have occurred just by random chance. The lower the p-value, the greater the statistical significance of the observed difference. P-value can be used as an alternative to or in a addition to pre-selected confidence levels for hypothesis testing [2].

- degrees of freedom: for the chi-square are calculated using the following formula: df = (r-1)(c-1) where r is the number of rows and c is the number of columns. If the observed chi-square test statistic is greater than the critical value, the null hypothesis can be rejected [3]

- expected frequencies: formula for expected frequencies is "Expected Frequency = (Row Total * Column Total)/N." 
The expected frequency is a probability count that appears in contingency table calculations including the chi-square test. [4]

- Null hypothesis: The null hypothesis of the Chi-Square test is that no relationship exists on the categorical variables in the population; they are independent. In this case the null hypothesis is that each person's neighborhood of residence (A, B, C, D) is independent of the person's occupational classification (white collar, blue collar, no collar). In this task the null hypothisis is: Each person's neighborhood of residence is independent of the person's occupational classification. 



In [11]:
# Code adjusted from: https://machinelearningmastery.com/chi-squared-test-for-machine-learning/   [5]

# ss. chi2_contingency(total) - this function takes and array as input (total) representing the contingency 
# table for the two or categorical variables. It returns calculated statistic and p-value for interpretation
# and degrees of freedom and table of expected freguencies. 
stat, p, dof, expected = ss.chi2_contingency(total)
print('dof=%d' % dof)
print(expected)
# probability of 95% can be used, suggesting that the finding of the test is quite likely given the assumption of
# the test that the variable is independent. If the statistic is less than or equal to the critical value, we can fail to reject
# this assumption, otherwise it can be rejected.
prob = 0.95
critical = ss.chi2.ppf(prob, dof)
print('probability=%.5f, critical=%.5f, stat=%.5f' % (prob, critical, stat))
if abs(stat) >= critical:
	print('Dependent (reject H0)')
else:
	print('Independent (fail to reject H0)')
# We can also interpret the p-value by comparing it to a chosen significance level, which would be 5%, calculated 
# by inverting the 95% probability used in the critical value interpretation.
alpha = 1.0 - prob
print('significance=%.5f, p=%.5f' % (alpha, p))
if p <= alpha:
	print('Dependent (reject H0)')
else:
	print('Independent (fail to reject H0)')

dof=6
[[ 80.53846154  80.53846154 107.38461538  80.53846154]
 [ 34.84615385  34.84615385  46.46153846  34.84615385]
 [ 34.61538462  34.61538462  46.15384615  34.61538462]]
probability=0.95000, critical=12.59159, stat=24.57120
Dependent (reject H0)
significance=0.05000, p=0.00041
Dependent (reject H0)


##### 2.3. Analysis and conclusion


The aim of this task is to verify Chi sqared value using scipy.stats and calculate p-value. Data that I have got shows the same Chi sqared value as the one that is shown in the given example on wikipedia page. In both cases Chi square is 24.6. To be more precise I got the result of 24.57120 with p-value of 0.00041. P-value tells us if we need to reject null hypothesis or not. Null hypothesis is an assumption that there is no relationship between variables. In this case four different neighbourhoods (A, B, C, D) were tested to see if there is any connection between neighbourhoods and person's occupational profession (white collar, blue collar, no collar). As p-value is 0.00041 it tells us that we need to reject null hypothesis (H0) which means that there is significant relationship between the neighbourhood and the occupation of people who live there. In order to show significant difference between the variables p-value should be less than 0.05 if probability is defined in 95%. 

To conclude, Chi sqared test is extremely important in hypothesis testing. In order to get accurate results aims, null hypothesis and problems of the research should be clearly defined before the actual testing. It is a great starting method in a research since it shows if there is any statistical difference between the variables and gives us valuable insight into the core of the problems. From my point of view, Chi sqared test should be done when we have done our research of the topic and if the test shows there is significant difference between tested variables, we shouldn't stop there, we should keep researching and asking more questions in order to find out what the real problems are. Numbers that we get using Chi sqared test don't really mean much without understanding the causes and consequences of the problems or variables that we are testing. 



##### 2.4. References

1. Chi squared test:  https://www.statisticssolutions.com/using-chi-square-statistic-in-research/
2. P-value: https://www.investopedia.com/terms/p/p-value.asp#:~:text=A%20p%2Dvalue%20is%20a,confidence%20levels%20for%20hypothesis%20testing
3. Degrees of freedom: https://www.statisticssolutions.com/non-parametric-analysis-chi-square/#:~:text=The%20degrees%20of%20freedom%20for,null%20hypothesis%20can%20be%20rejected.
4. Expected frequencies: https://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/BS704_HypothesisTesting-ChiSquare/BS704_HypothesisTesting-ChiSquare_print.html#:~:text=Expected%20Frequency%20%3D%20
5. Chi square test code: https://machinelearningmastery.com/chi-squared-test-for-machine-learning/
6. Chi squared test: https://en.wikipedia.org/wiki/Chi-squared_test